File: src/ciphertext.erl
-module(ciphertext).

-export([
    %% Ciphertext API
    verify/1,
    cmp/2,
    serialize/1,
    deserialize/1
]).

-type ciphertext() :: reference().
-export_type([ciphertext/0]).

-spec verify(Ciphertext :: ciphertext()) -> boolean().
verify(Ciphertext) ->
    erlang_tc:ciphertext_verify(Ciphertext).

-spec cmp(C1 :: ciphertext(), C2 :: ciphertext()) -> boolean().
cmp(C1, C2) ->
    erlang_tc:ciphertext_cmp(C1, C2).

-spec serialize(Ciphertext :: ciphertext()) -> binary().
serialize(Ciphertext) ->
    erlang_tc:ciphertext_serialize(Ciphertext).

-spec deserialize(BinC :: binary()) -> ciphertext().
deserialize(BinC) ->
    erlang_tc:ciphertext_deserialize(BinC).

File: native/ciphertext.rs
use crate::bin::Bin;
use rustler::{Env, ResourceArc};
use serde::{Deserialize, Serialize};
use threshold_crypto::Ciphertext;

/// Struct to hold PublicKeyShare
#[derive(Serialize, Deserialize)]
pub struct CiphertextRes(pub(crate) Ciphertext);

pub type CiphertextArc = ResourceArc<CiphertextRes>;

pub fn load(env: Env) -> bool {
    rustler::resource!(CiphertextRes, env);
    true
}

#[rustler::nif(name = "ciphertext_verify")]
fn ciphertext_verify(cipher: CiphertextArc) -> bool {
    cipher.0.verify()
}

#[rustler::nif(name = "ciphertext_cmp")]
fn ciphertext_cmp(c1a: CiphertextArc, c2a: CiphertextArc) -> bool {
    c1a.0 == c2a.0
}

#[rustler::nif(name = "ciphertext_serialize")]
pub fn ciphertext_serialize(ca: CiphertextArc) -> Bin {
    // TODO: Investigate allowing specifying encoding type using an erlang atom
    let bytes = bincode::serialize(&ca.0).unwrap();
    Bin(bytes)
}

#[rustler::nif(name = "ciphertext_deserialize")]
pub fn ciphertext_deserialize(bin: rustler::Binary) -> CiphertextArc {
    CiphertextArc::new(bincode::deserialize(&bin).unwrap())
}
